runtime/src/bank.rs

```diff
@@ -2166,97 +2172,100 @@ impl Bank {
         &self,
         thread_pool: &ThreadPool,
         reward_calc_tracer: Option<impl Fn(&RewardCalculationEvent) + Send + Sync>,
-    ) -> DashMap<Pubkey, VoteWithStakeDelegations> {
+    ) -> LoadVoteAndStakeAccountsResult {
         let stakes = self.stakes.read().unwrap();
-        let accounts = DashMap::with_capacity(stakes.vote_accounts().as_ref().len());
+        let vote_with_stake_delegations_map =
+            DashMap::with_capacity(stakes.vote_accounts().as_ref().len());
+        let invalid_stake_keys_set: DashSet<Pubkey> = DashSet::new();
+        let invalid_vote_keys_set: DashSet<Pubkey> = DashSet::new();
 
         thread_pool.install(|| {
             stakes
                 .stake_delegations()
                 .par_iter()
                 .for_each(|(stake_pubkey, delegation)| {
                     let vote_pubkey = &delegation.voter_pubkey;
-                    let stake_account = match self.get_account_with_fixed_root(stake_pubkey) {
-                        Some(stake_account) => stake_account,
-                        None => return,
-                    };
+                    if invalid_vote_keys_set.contains(vote_pubkey) {
+                        return;
+                    }
 
-                    // fetch vote account from stakes cache if it hasn't been cached locally
-                    let fetched_vote_account = if !accounts.contains_key(vote_pubkey) {
-                        let vote_account = match self.get_account_with_fixed_root(vote_pubkey) {
-                            Some(vote_account) => vote_account,
-                            None => return,
-                        };
+                    let stake_delegation = match self.get_account_with_fixed_root(stake_pubkey) {
+                        Some(stake_account) => {
+                            if stake_account.owner() != &solana_stake_program::id() {
```

Comment:
Yes, they must be but there is a chance that a stake account was deleted and the account now belongs to a different program without the cache catching the change in ownership. That should be patched up now that we have the following feature activated:
```
HFpdDDNQjvcXnXKec697HDDsyk6tFoWS2o8fkxuhQZpL | active since slot 110592000 | remove delegations from stakes cache when inactive
```
But if we failed to remove inactive delegations before that feature was activated, they could still be stuck in the cache. I would say that after we do a final prune (which this PR adds a feature for), it should be safe to add assertions to the cache.

Reply:
Makes sense, thanks!